### PR TITLE
Implement revision content format and evolution operations

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Summary

This PR implements the `revision` content format for representing mutable object evolution on top of an immutable content-addressable store, along with core store-touching operations for head resolution and content materialization.

## Key Changes

- **New `fs/media/revision/` module**: Defines the pure format specification
  - `module.f.ts`: RTTI schema for revisions with fields for object identity, parent links, content/changes, generation tracking, and archival status
  - `README.md`: Complete spec covering the format, materialization algorithm, versioning rules, and design rationale
  - `proof.f.ts`: Comprehensive tests for schema validation and semantic constraints (hash-only parents, ref validation, etc.)

- **New `fs/cas/evo/` module**: Store-touching operations for the revision format
  - `module.f.ts`: Core operations including:
    - `readRevision`: Safely decodes blobs as revisions, handling missing/invalid/malformed content
    - `revisionsOf`: Finds all revisions of a given object
    - `heads`: Resolves current head(s) via reverse-index of parent references
    - `materialize`: Resolves content with precedence (explicit content > parent recursion > object identity)
    - `computeGeneration`: Recomputes generation from parent chain to detect cache mismatches
  - `proof.f.ts`: Extensive test suite covering linear history, branching/merging, head demotion on sync, materialization edge cases, and error conditions

- **New `fs/cas/README.md`**: High-level documentation of the CAS module and its submodules

- **Refactored `fs/cas/module.f.ts`**: Extracted `collectRead` helper (drains read streams into single Vec) for reuse in both MCP and evo modules

- **Updated `fs/cas/mcp/module.f.ts`**: Removed duplicate `collectRead` implementation, now imports from core module

- **Updated `fs/cas/todo/revision-content-format.md`**: Marked completed tasks for format spec, schema, and core operations

## Notable Implementation Details

- **Semantic validation beyond schema**: `decodeRevision` validates not just shape but also that parents are hashes (never bridge URLs) and all refs are recognized forms (cbase32 or https://)
- **Acyclic DAG by construction**: Parent references use content hashes, which are only known after the parent blob exists, preventing cycles without explicit guards
- **Scoped head resolution**: A revision of object B referencing object A's hash never demotes A's heads—the reverse index is per-object
- **First iteration limitations**: Materialization supports zero or one parent and no `changes` field; multi-parent merges must carry explicit `content`
- **Error propagation**: `computeGeneration` propagates parent decode failures rather than treating corrupt ancestors as generation 0

https://claude.ai/code/session_01JjRZPdUUnto7YbG3eH63gq

### CODE REVIEW

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/evo/module.f.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import { error, ok, type Result } from '../../types/result/module.f.ts'
+import { foldStep, okStep, pure, type Effect, type Operation } from '../../effects/module.f.ts'
+import { collectRead, type Cas } from '../module.f.ts'
+import type { Vec } from '../../types/bit_vec/module.f.ts'
+import { toOption } from '../../types/nullable/module.f.ts'
+
+/** Parses `text` as JSON, reporting a syntax error as a `Result` instead of throwing. */
+const tryParseJson = (text: string): Result<Unknown, string> => {
+    try {
+        return ok(parseJson(text))
```
Comment (sergey-shandar) [inline fs/cas/evo/module.f.ts]: Don't use `try cacth`. Use `fs/media/json/parser/..` instead.

File: fs/cas/evo/proof.f.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+import { vec, vec8, type Vec } from '../../types/bit_vec/module.f.ts'
+import { utf8 } from '../../text/module.f.ts'
+import { ok, unwrap, type Ok, type Result } from '../../types/result/module.f.ts'
+import type { IoResult } from '../../effects/node/module.f.ts'
+import { mimeType } from '../../media/revision/module.f.ts'
+import { heads, materialize, computeGeneration, readRevision, revisionsOf } from './module.f.ts'
+
+/** `Result` values are tuples, so `assertEq`'s `===` never matches two
+ *  independently constructed ones — compare tag and payload separately. */
+const assertOk = <T>(r: Result<T, unknown>, expected: T): void => {
```
Comment (sergey-shandar) [inline fs/cas/evo/proof.f.ts]: use destruction `[t, v]` instead of `r`

File: fs/cas/evo/proof.f.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        run(effect)
+    },
+
+    // Branch + merge: two revisions branch off a common first revision, then
+    // a merge revision (two parents, `content`) resolves both into one head.
+    branchAndMerge: () => {
+        const effect =
+            writeJson({ mimeType, object: objectA, parents: [], content: 'https://example.com/root' }).step(hashRoot => {
+                const root = vecToCBase32(hashRoot)
+                return writeJson({ mimeType, object: objectA, parents: [root], content: 'https://example.com/left' }).step(hashLeft => {
```
Comment (sergey-shandar) [inline fs/cas/evo/proof.f.ts]: Don't do nesting `step`. Write 
```ts
operation(...x)
    .step(x => operation(...x))
    .step(x => operation(...x))
```
instead of 
```ts
operation(...x)
    .step(x => operation(...x)
        .step(x => operation(...x)))
```

File: fs/cas/evo/proof.f.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                })
+            })
+        run(effect)
+    },
+
+    // Many heads: three independent first revisions of the same object are
+    // all heads at once (no merge yet).
+    manyHeadsForOneObject: () => {
+        const effect =
+            writeJson({ mimeType, object: objectA, parents: [], content: 'https://example.com/one' }).step(() =>
```
Comment (sergey-shandar) [inline fs/cas/evo/proof.f.ts]: same! No nesting!

File: fs/media/revision/module.f.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+export const hash: RttiString = string
+
+/** True when `s` is a valid cbase32-encoded CAS hash. */
+export const isHash = (s: string): boolean => cBase32ToVec(s) !== null
+
+/** True when `s` is a `https://` bridge URL. */
+export const isHttpsRef = (s: string): boolean => s.startsWith('https://')
+
+/** True when `s` is a recognized `ref`: a cbase32 hash or a `https://` URL. */
+export const isRef = (s: string): boolean => isHash(s) || isHttpsRef(s)
```
Comment (sergey-shandar) [inline fs/media/revision/module.f.ts]: No need to handle `isHttpsRef` any more, we decided to use mimeType for better detecting a type of a content.

File: fs/media/revision/module.f.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+ * bridge URL cannot stand in for it. Schema-wise identical to `ref`;
+ * `isHash` performs the semantic check.
+ */
+export const hash: RttiString = string
+
+/** True when `s` is a valid cbase32-encoded CAS hash. */
+export const isHash = (s: string): boolean => cBase32ToVec(s) !== null
+
+/** True when `s` is a `https://` bridge URL. */
+export const isHttpsRef = (s: string): boolean => s.startsWith('https://')
```
Comment (sergey-shandar) [inline fs/media/revision/module.f.ts]: no need

File: fs/media/revision/module.f.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+const validateRevision = validate(revision)
+
+/**
+ * Decodes an untrusted value as a `Revision`, validating both the shape
+ * (via the rtti schema) and the semantic constraints the schema alone can't
+ * express: every `parents` entry must be a hash (never a bridge URL), and
+ * `object` / `content` / `changes` entries must be recognized `ref`s.
+ */
+export const decodeRevision = (value: Unknown): Result<Revision, string> => {
```
Comment (sergey-shandar) [inline fs/media/revision/module.f.ts]: use only valid RTTI parsing, no need for this complications.

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

File: fs/media/revision/module.f.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+     * Marks the mutable object as archived/inactive. The revision still
+     * exists; it is just hidden from normal active views.
+     */
+    archived: option(true),
+} as const
+
+export type Revision = Ts<typeof revision>
+
+/** Decodes an untrusted value as a `Revision` — plain rtti shape validation. */
+export const decodeRevision = validate(revision)
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/media/revision/module.f.ts:77]: This is intentional, per an earlier review round on this same PR (see the "use only valid RTTI parsing, no need for this complications" thread above) — `decodeRevision` is deliberately kept to plain RTTI shape validation. Ref/hash format checking is deferred to the point of use: `fs/cas/evo`'s `toVec`/`cBase32ToVec` already drops an invalid `parents` entry gracefully, and a `materialize`/`heads` caller resolving `object`/`content` against the store gets the same graceful failure there rather than a duplicate check here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/evo/module.f.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * Resolves a `parents` hash string to its `Vec`. `fs/media/revision` does
+ * not itself check that a `parents` entry decodes to a valid hash, so an
+ * invalid one is simply dropped here (`toOption`) rather than treated as an
+ * error — a store that only ever writes through `fs/media/revision`'s
+ * schema can still write a `parents` string that isn't a real hash, and
+ * silently ignoring it is the same graceful handling an absent parent gets.
+ */
+const toVec = (s: string): readonly Vec[] => toOption(cBase32ToVec(s))
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/evo/module.f.ts:70]: Same design decision as the sibling thread on `fs/media/revision/module.f.ts:77` (see above — "use only valid RTTI parsing, no need for this complications") — `toVec`'s doc comment already calls this out explicitly: dropping an unresolvable `parents` entry is treated the same as an absent parent, not as an error. Adding a distinct "malformed vs. absent" error path here would be exactly the semantic ref-validation layer that was intentionally left out.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fs/cas/evo/module.f.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+ * Fails when: the blob does not decode as a `Revision`; it carries
+ * `changes` (unimplemented); it has more than one parent without `content`
+ * (a merge revision, per the spec, must carry `content`); or its single
+ * parent is a revision of a different `object`.
+ */
+export const materialize = <O extends Operation>(c: Cas<O>) => (hash: Vec): Effect<O, Result<string, string>> =>
+    readRevision(c)(hash).step(okStep(revision => materializeRevision(c)(revision)))
+
+const materializeRevision = <O extends Operation>(c: Cas<O>) => (revision: Revision): Effect<O, Result<string, string>> => {
+    if (revision.content !== undefined) { return pure(ok(revision.content)) }
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/evo/module.f.ts:121]: Same design decision as the `fs/media/revision/module.f.ts:77` thread above — `materialize` returns whatever `content` the schema says is present without re-validating its ref format; a caller resolving that ref against the store gets the same graceful failure `fs/cas/evo` already gives an invalid `parents` hash, rather than a duplicate check here.

---
_Generated by [Claude Code](https://claude.ai/code)_
